control-plane/pkg/reconciler/consumer/consumer.go

```diff
@@ -0,0 +1,409 @@
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get %s: %w", kafkainternals.ConsumerGroupGroupVersionKind.Kind, err)
+	}
+	if apierrors.IsNotFound(err) {
+		return nil, nil
+	}
+
+	// Get user facing resource from the owner references list.
+	var cgOr *metav1.OwnerReference
+	for i, or := range cg.OwnerReferences {
+		k := strings.ToLower(or.Kind)
+		// TODO hardcoded resource kinds.
+		if k == "trigger" || k == "kafkasource" || k == "kafkachannel" {
+			cgOr = &cg.OwnerReferences[i]
+			break
+		}
+	}
+
+	userFacingResource := cgOr
+	ref := &contract.Reference{
+		Uuid:      string(userFacingResource.UID),
+		Namespace: c.GetNamespace(),
+		Name:      userFacingResource.Name,
+	}
+	return ref, nil
+}
+
+func reconcileDeliveryOrder(c *kafkainternals.Consumer) contract.DeliveryOrder {
+	if c.Spec.Delivery == nil {
+		return contract.DeliveryOrder_UNORDERED
+	}
+	switch c.Spec.Delivery.Ordering {
+	case eventing.Ordered:
+		return contract.DeliveryOrder_ORDERED
+	case eventing.Unordered:
+		return contract.DeliveryOrder_UNORDERED
+	}
+	return contract.DeliveryOrder_UNORDERED
+}
+
+func (r Reconciler) reconcileReplyStrategy(ctx context.Context, c *kafkainternals.Consumer, egress *contract.Egress) error {
+	if c.Spec.Reply == nil {
+		return nil
+	}
+	if c.Spec.Reply.NoReply != nil {
+		egress.ReplyStrategy = &contract.Egress_DiscardReply{}
+		return nil
+	}
+	if c.Spec.Reply.URLReply != nil {
+		destination, err := r.Resolver.URIFromDestinationV1(ctx, c.Spec.Reply.URLReply.Destination, c)
+		if err != nil {
+			return fmt.Errorf("failed to resolve reply destination: %w", err)
+		}
+		egress.ReplyStrategy = &contract.Egress_ReplyUrl{
+			ReplyUrl: destination.String(),
+		}
+		return nil
+	}
+	if c.Spec.Reply.TopicReply != nil {
+		egress.ReplyStrategy = &contract.Egress_ReplyToOriginalTopic{}
+		return nil
+	}
+
+	return nil
+}
+
+func reconcileFilters(c *kafkainternals.Consumer) *contract.Filter {
+	if c.Spec.Filters == nil {
+		return nil
+	}
+	if c.Spec.Filters.Filter != nil {
+		return &contract.Filter{Attributes: c.Spec.Filters.Filter.Attributes}
+	}
+
+	return nil
+}
+
+type contractMutatorFunc func(logger *zap.Logger, ct *contract.Contract, c *kafkainternals.Consumer) int
+
+func addResource(resource *contract.Resource) contractMutatorFunc {
+	return func(logger *zap.Logger, ct *contract.Contract, c *kafkainternals.Consumer) int {
+		return coreconfig.AddOrUpdateResourceConfig(ct, resource, coreconfig.FindResource(ct, c.GetUID()), logger)
+	}
+}
+
+func removeResource(_ *zap.Logger, ct *contract.Contract, c *kafkainternals.Consumer) int {
+	idx := coreconfig.FindResource(ct, c.GetUID())
+	if idx == coreconfig.NoResource {
+		return coreconfig.NoResource
+	}
+	coreconfig.DeleteResource(ct, idx)
+	return idx
+}
+
+func (r Reconciler) schedule(ctx context.Context, logger *zap.Logger, c *kafkainternals.Consumer, mutatorFunc contractMutatorFunc) (bool, error) {
```

Comment:
Can you explain what the `schedule` function does in terms of figuring out placements here? At this point, the consumer is already scheduled right? Maybe some more comments here might help as in what is the update to the config map being done, etc?

Reply:
I'll add comment, maybe `schedule` as the name isn't the best one.

Reply:
Does this help?
```golang
// schedule mutates the ConfigMap associated with the pod specified by Consumer.Spec.PodBind.
//
// The actual mutation is done by calling the provided contractMutatorFunc.
```

Reply:
It does. Great. Think the function being called `schedule` made it a bit confusing without a comment